rfd/0195-vnet-windows.md

```diff
@@ -0,0 +1,339 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 195 - Windows VNet
+
+## Required Approvers
+
+* Engineering: @ravicious && (@zmb3 || @rosstimothy)
+* Security: doyensec
+
+## What
+
+This document outlines the design for [VNet](0163-vnet.md) on Windows clients.
+
+## Why
+
+The design for Windows differs significantly from VNet on MacOS.
+On MacOS we use a Launch Daemon running as root to create a virtual TUN network
+interface, and then pass the TUN to a client process (Connect or tsh) over a pipe.
+The client process then manages all networking and the daemon handles OS
+configuration.
+
+On Windows, there is no way to pass the TUN interface over a pipe, so all
+networking is handled in a Windows Service, with IPC to the user process which
+handles all Teleport client methods.
+The client process retains control of all Teleport client methods and user keys
+so that it can easily perform MFA prompts and hardware key signatures using
+existing code.
+
+## Details
+
+### UX
+
+The goal is to keep the VNet UX identical on Windows and MacOS.
+Teleport Connect will have support VNet with the best UX for details like MFA
+prompts and error reporting.
+`tsh vnet` will be supported on Windows just as it is on MacOS, but building
+a first-class CLI UX will not be a goal, Connect will be the recommended client.
+
+In keeping with the MacOS experience, this design avoids administrator (UAC)
+prompts each time VNet is started, only requiring a prompt the very first time
+VNet is started on a Windows client machine.
+Errors and any user prompts will be identical on each OS, with a goal of reusing
+as much code between operating systems as possible so that this happens by
+default.
+
+### Virtual Network Interface
+
+VNet requires a virtual network interface to be created on the host OS,
+the specific kind of interface we use is typically called a TUN interface.
+VNet configures IP routes on the host so that IP traffic for a set of configured
+CIDR ranges are bound to the TUN.
+VNet then reads and writes IP packets to and from the TUN interface to handle
+TCP connections to Teleport apps and UDP connections to VNet's internal DNS server.
+
+Because Windows does not natively support TUN interfaces the same way that MacOS
+and Linux do, we will leverage an open-source driver wintun.dll to provide the
+TUN interface.
+The DLL is available for download from https://www.wintun.net/ as a signed DLL.
+It has a custom license that permits commercial use as long as we use their
+signed DLL via the provided API.
+Another open-source library that we already use on MacOS provides a common Go
+interface for the TUN, this is `golang.zx2c4.com/wireguard/tun`.
+
+The signed `wintun.dll` file will be distributed with Connect and installed in
+the same directory as `tshd`.
```

Comment:
As I mentioned in the other comment, the directory that has `tsh.exe` is added to Path. https://github.com/gravitational/teleport/blob/master/rfd/0099-link-bundled-tsh.md

From what I see, this shouldn't pose a problem to us specifically. It might be a problem for other apps which depend on the same dll being in Path, but I'm not that familiar with best practices there. https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-unpackaged-apps

Reply:
in case multiple versions of wintun are installed on the system, i will copy it to the same directly where `tsh` is installed for the windows service, so that the right version takes precedence